### DISCUSSION TITLE
Not sure how to train TTS model

### DISCUSSION THREAD

**Discussion by davidmartinrius** [NVIDIA-NeMo/NeMo#4305] [Q&A] [answered]
Hello!

In this post I would like to ask some questions.
As a premise I want to tell you what kind of machine I am working.

I have got an MSI Creator P100X -637
The specs are 
GPU RTX 3080 
64GB RAM
Intel Core i7-11700K
CPU 3.6 GHz

I work with trial an error and it is very confusing. Depending on the size of the dataset the GPU only accepts a limited max number in batch size.

For example, if I am going to train a dataset of 20 hours it will only accept a max batch size of 6. If I try a greater number it always raises an CUDA out of memory. (I supose it is the most common error when training)

On the other hand, if I am going to train a dataset of 5 hours aprox I can work with a batch size of 14 or 15.

### **Question 1/4.**
Is there a way to estimate the batch size previous to train?

Each time that I start training, if there is too much batch size it will crash. From I start training to crash it can pass like 15-20 minutes. This is because before start training the program needs to load all data previously. So  i wasted a lot of time because until I don't find the good max batch size I can spend hours.

### **Question 2/4.**
Is there a way to estimate how many epochs to train a model based on the size of the dataset and the GPU batch size?

Now I set like 1000 epochs and I let the model training until the training loss is near to 0. But I am sure there is a cleaner way to do it.

### **Question 3/4.**
How to know how much time to train a model?
I supervised the trainings in tensorboard.
I deduced that I should evaluate the training loss of the model to know if is enough trained. But I am not sure of this.

### **Question 4/4.**
How to evaluate when a model is enough trained?

I spend a lots of hours trying to train a model without success. I am sure a lot of people that start with NeMo has this kind of doubts. Also  I read a lot of discussions, issues, etc but I still did not solve this doubts.

So I would apreciate some help. I really want to train a TTS model with my own datasets.

Thank you!

**Comment by ericharper** [comment]
@redoctopus could you take a look at this. I will convert it to a discussion.

**Comment by redoctopus** [comment, marked as answer]
These seem to be general ML model training questions rather than NeMo-specific or even TTS-specific ones--I can give you some rules of thumb but would also recommend finding some general resources or classes as well.

1. Maximum batch size depends on the amount of memory your GPU has, as well the size of individual samples. The total size of your dataset shouldn't matter as much as the memory footprint of the batches, which for TTS depends on audio length of individual entries; to this end we usually recommend setting a maximum audio length if you have a few very long samples. You can estimate max batch size yourself by knowing how much memory your GPU has free (based on the size of your model), roughly divided by the largest input size. Though to be honest, mostly I guess-and-check by guessing some smaller initial value, eyeballing memory usage, then estimating how much I can scale up batch size.
2. We have some roughly reasonable defaults in the configs, but this really depends on the dataset and the model. Generally, larger models will take longer, but smaller datasets will take less time. At this most general level, my best advice is to keep an eye on your training and validation loss.
3. This also depends on your hardware, model, dataset, etc. You can get a good idea of how long it'll take by checking against published results. Papers in which a model is introduced will usually include time taken to train on specific hardware and a specific dataset, and you can use this as a comparison to your setup.
4. This is a very broad question; for TTS specifically this is somewhat more difficult, but beyond tracking loss, you can also add predicted and ground truth spectrograms (where applicable) to the tensorboard logger to keep an eye on progress and make sure the model is progressing.

Best of luck!
